### DISCUSSION TITLE
Clarification on DicomClient API semantics

### DISCUSSION THREAD

**Discussion by ayedo** [fo-dicom/fo-dicom#1639] [Q&A]
First of all, thanks for your great work on fo-dicom :)

I'm confused about the API semantics of `IDicomClient`. The way I understand is that, you can add multiple requests and then you call Send. My questions are: 

- when the Send call returns, does that mean that all the requests that were added were successfully sent?
- are you supposed to add more requests to the same client after calling Send?

Thanks, and have a nice day!

Alon

**Comment by ayedo** [comment]
I just found this in the source code: 

![image](https://github.com/fo-dicom/fo-dicom/assets/4739734/9580dae2-1d9c-4359-ae3f-df063e05463f)

So, that means the DicomClient is kind of "single use" and cannot be reused after calling `SendAsync`?

**Comment by gofal** [comment]
when the DicomClient is opening an connection to the server, then one part of this handshake is the AssociationRequest, where the client has to tell the server, which kind of data or requests the client is about to send and the server can response which message are supported and which not. 
so it can introduce some critical errors if you would add some requests after the association has been opened. This is why you should first add all the requests you want to send, and then with the method SendAsync you actually start the sending.

It might work, if you then reuse the instance, add more Requests and then call the SendAsync again, but it is not recomended.

If you want to have full control, you should use the AdvancedClient. There you open an association manually and you have to set all the PresentationContexts that you want to send. and then you can send any request that you want in "real time".
